src/graph/src/graph.ts

```diff
@@ -40,8 +42,18 @@ const getResolutionCacheKey = (
 ): string => {
   const f = spec.final
   // if it's a file: dep, then the fromNode location matters
-  const fromPref = f.type === 'file' ? location + ' : ' : ''
-  return fromPref + String(f) + queryModifier
+  const fromPrefix = f.type === 'file' ? location + ' : ' : ''
+  // the unique key should also precise what is the type of the spec,
+  // and in the case of a registry, what registry it is from.
+  const typePrecisionKey =
+    f.registry ?
+      `${cacheKeySeparator}registry` +
+      `${cacheKeySeparator}${f.registry}`
+    : f.gitRemote ?
+      `${cacheKeySeparator}git` + `${cacheKeySeparator}${f.gitRemote}`
+    : `${cacheKeySeparator}${f.type}`
+  const modifierSuffix = `${cacheKeySeparator}${queryModifier}`
+  return fromPrefix + String(f) + typePrecisionKey + modifierSuffix
```

Comment:
these cache keys needed to be aware of more information such as registries so that it won't just retrieve a package with the same name that came from a different registry